Source/JavaScriptCore/runtime/ISO8601.cpp

```diff
@@ -70,50 +70,37 @@ static int32_t parseDecimalInt32(const CharType* characters, unsigned length)
 // https://tc39.es/proposal-temporal/#sec-temporal-durationhandlefractions
 static void handleFraction(Duration& duration, int factor, StringView fractionString, TemporalUnit fractionType)
 {
-    ASSERT(fractionString.length() && fractionString.length() <= 9 && fractionString.isAllASCII());
+    auto fractionLength = fractionString.length();
+    ASSERT(fractionLength && fractionLength <= 9 && fractionString.isAllASCII());
     ASSERT(fractionType == TemporalUnit::Hour || fractionType == TemporalUnit::Minute || fractionType == TemporalUnit::Second);
 
-    if (fractionType == TemporalUnit::Second) {
-        Vector<LChar, 9> padded(9, '0');
-        for (unsigned i = 0; i < fractionString.length(); i++)
-            padded[i] = fractionString[i];
-        duration.setMilliseconds(factor * parseDecimalInt32(padded.data(), 3));
-        duration.setMicroseconds(factor * parseDecimalInt32(padded.data() + 3, 3));
-        duration.setNanoseconds(factor * parseDecimalInt32(padded.data() + 6, 3));
-        return;
-    }
+    Vector<LChar, 9> padded(9, '0');
+    for (unsigned i = 0; i < fractionLength; i++)
+        padded[i] = fractionString[i];
 
-    double fraction = factor * parseInt(fractionString, 10) / std::pow(10, fractionString.length());
+    int64_t fraction = factor * parseDecimalInt32(padded.data(), 9);
```

Comment:
Let's cast `factor` or parseDecimalInt32 result with `static_cast<int64_t>(...)` to ensure that multiplication is done in int64_t.

Reply:
Oh, good catch!